File: test/rust/crates/specgate-ctsc-fixtures/src/conformance/structured/scalar_types.rs
use specgate::{SpecEvent, spec_operation};

#[derive(Debug, Clone, PartialEq, Eq, SpecEvent)]
#[spec_component("fixture.scalar_types")]
pub enum ScalarValue {
    Integer(i64),
    Boolean(bool),
}

#[spec_operation("classify", spec = "fixture.scalar_types")]
pub fn classify(id: i64, active: bool) -> ScalarValue {
    if active {
        ScalarValue::Integer(id)
    } else {
        ScalarValue::Boolean(false)
    }
}

#[test]
fn i64_and_bool_scalars_are_capturable() {
    assert_eq!(
        classify(9_000_000_000, true),
        ScalarValue::Integer(9_000_000_000)
    );
    assert_eq!(classify(9_000_000_000, false), ScalarValue::Boolean(false));
}

File: test/csharp/SpecGate.CtscFixtures/Conformance/Structured/ScalarTypes.cs
using SpecGate.Annotations;

namespace SpecGate.CtscFixtures.Conformance.Structured;

[SpecEvent("ScalarValue")]
public abstract class ScalarValue;

[SpecEvent("Integer")]
public sealed class IntegerValue : ScalarValue
{
    [SpecEvent("value")]
    public long Value { get; set; }
}

[SpecEvent("Boolean")]
public sealed class BooleanValue : ScalarValue
{
    [SpecEvent("value")]
    public bool Value { get; set; }
}

public static class ScalarTypes
{
    [SpecOperation("classify", Spec = "fixture.scalar_types")]
    public static ScalarValue Classify(
        [SpecInput("id")] long id,
        [SpecInput("active")] bool active) =>
        active ? new IntegerValue { Value = id } : new BooleanValue();
}
